modules/dxp/apps/multi-factor-authentication/multi-factor-authentication-fido2-web/src/main/resources/META-INF/resources/js/AuthenticationTransformer.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * The contents of this file are subject to the terms of the Liferay Enterprise
+ * Subscription License ("License"). You may not use this file except in
+ * compliance with the License. You can obtain a copy of the License by
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+import * as webauthn from './webauthn';
+
+export default function AuthenticationTransformer({
+	additionalProps,
+	portletNamespace,
+	...otherProps
+}) {
+	return {
+		...otherProps,
+		onClick() {
+			let {assertionRequest} = additionalProps;
+
+			if (!assertionRequest) {
+				return;
+			}
+
+			assertionRequest = JSON.parse(assertionRequest);
+
+			webauthn
+				.get(assertionRequest.publicKeyCredentialRequestOptions)
+				.then((value) => {
+					const publicKey = webauthn.credentialToObject(value);
+
+					if (
+						publicKey.response.userHandle !== null &&
+						publicKey.response.userHandle === ''
+					) {
+						delete publicKey.response.userHandle;
+					}
+
+					const responseInput = document.getElementById(
+						`${portletNamespace}responseJSON`
+					);
+
+					responseInput.value = JSON.stringify(publicKey);
```

Comment:
> Anyone know if we have any sort of shared "global state" across js components?

We don't exactly. You can do something with `Liferay.on` and `Liferay.fire` to wire up ad hoc communication between disparate places, or roll your own mechanism; for example:

- "Hiding" something globally with `Symbol.for()`: `window[Symbol.for('private stuff')] = 1`.
- Sticking something in [`sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage).
- etc

We have these open issues:

- [liferay-frontend-guidelines#44](https://github.com/liferay/liferay-frontend-guidelines/issues/44) ("Standardize state management with React"); note this one is just about local state within a React app.
- [LPS-98395](https://issues.liferay.com/browse/LPS-98395) ("Client Side State Management in Liferay Portal"); this one is actually about cross-widget shared state.

Reply:
@bryceosterhaus , @wincent thanks for those comments
I think that for the moment I'm gonna leave it as is, but if this can be improved in the future lets.

Reply:
@julien I agree, I think using `.on` and `.fire` would be overkill for this.